File: server/src/collectors/suumoRentalDensity.js
/**
 * Suumo — rental listing density per prefecture.
 *
 * Suumo's rental search at https://suumo.jp/chintai/ encodes prefecture
 * in the URL (`/chintai/{prefecture-code}/`) and each search page shows
 * a total count. We fetch just the first page for each prefecture and
 * parse the listing total, pinning the result at the prefecture
 * centroid.
 *
 * LEGAL NOTE: Suumo's ToS is Japan-standard — casual light polling has
 * not been known to trigger blocks, but this is a scrape. Respect
 * robots.txt and keep the cadence >= 24h.
 *
 * No auth. Uses prefecture centroids pulled from the existing helper.
 */

const TIMEOUT_MS = 15000;

// Rough centroids of each prefecture. Used to pin the aggregate count
// somewhere sensible for map rendering.
const PREFECTURE_CENTROID = {
  '北海道': { lat: 43.0642, lon: 141.3469 }, '青森県': { lat: 40.8243, lon: 140.7400 },
  '岩手県': { lat: 39.7036, lon: 141.1527 }, '宮城県': { lat: 38.2688, lon: 140.8721 },
  '秋田県': { lat: 39.7186, lon: 140.1024 }, '山形県': { lat: 38.2404, lon: 140.3634 },
  '福島県': { lat: 37.7503, lon: 140.4676 }, '茨城県': { lat: 36.3418, lon: 140.4468 },
  '栃木県': { lat: 36.5657, lon: 139.8836 }, '群馬県': { lat: 36.3912, lon: 139.0608 },
  '埼玉県': { lat: 35.8570, lon: 139.6489 }, '千葉県': { lat: 35.6047, lon: 140.1233 },
  '東京都': { lat: 35.6895, lon: 139.6917 }, '神奈川県': { lat: 35.4478, lon: 139.6425 },
  '新潟県': { lat: 37.9161, lon: 139.0364 }, '富山県': { lat: 36.6953, lon: 137.2113 },
  '石川県': { lat: 36.5946, lon: 136.6256 }, '福井県': { lat: 36.0652, lon: 136.2216 },
  '山梨県': { lat: 35.6640, lon: 138.5683 }, '長野県': { lat: 36.6513, lon: 138.1811 },
  '岐阜県': { lat: 35.3912, lon: 136.7223 }, '静岡県': { lat: 34.9770, lon: 138.3830 },
  '愛知県': { lat: 35.1802, lon: 136.9066 }, '三重県': { lat: 34.7303, lon: 136.5086 },
  '滋賀県': { lat: 35.0045, lon: 135.8686 }, '京都府': { lat: 35.0212, lon: 135.7556 },
  '大阪府': { lat: 34.6863, lon: 135.5200 }, '兵庫県': { lat: 34.6913, lon: 135.1830 },
  '奈良県': { lat: 34.6851, lon: 135.8050 }, '和歌山県': { lat: 34.2261, lon: 135.1675 },
  '鳥取県': { lat: 35.5036, lon: 134.2378 }, '島根県': { lat: 35.4723, lon: 133.0505 },
  '岡山県': { lat: 34.6618, lon: 133.9350 }, '広島県': { lat: 34.3963, lon: 132.4596 },
  '山口県': { lat: 34.1859, lon: 131.4707 }, '徳島県': { lat: 34.0657, lon: 134.5592 },
  '香川県': { lat: 34.3401, lon: 134.0434 }, '愛媛県': { lat: 33.8416, lon: 132.7657 },
  '高知県': { lat: 33.5597, lon: 133.5311 }, '福岡県': { lat: 33.6064, lon: 130.4181 },
  '佐賀県': { lat: 33.2494, lon: 130.2988 }, '長崎県': { lat: 32.7503, lon: 129.8779 },
  '熊本県': { lat: 32.7898, lon: 130.7417 }, '大分県': { lat: 33.2382, lon: 131.6126 },
  '宮崎県': { lat: 31.9111, lon: 131.4239 }, '鹿児島県': { lat: 31.5602, lon: 130.5581 },
  '沖縄県': { lat: 26.2124, lon: 127.6809 },
};

// Suumo prefecture slugs. These are the two-letter kanji-romaji codes
// Suumo uses in the URL path (e.g. `/chintai/tokyo/`).
const SUUMO_SLUGS = {
  '北海道': 'hokkaido', '青森県': 'aomori', '岩手県': 'iwate', '宮城県': 'miyagi',
  '秋田県': 'akita', '山形県': 'yamagata', '福島県': 'fukushima',
  '茨城県': 'ibaraki', '栃木県': 'tochigi', '群馬県': 'gumma', '埼玉県': 'saitama',
  '千葉県': 'chiba', '東京都': 'tokyo', '神奈川県': 'kanagawa',
  '新潟県': 'nigata', '富山県': 'toyama', '石川県': 'ishikawa', '福井県': 'fukui',
  '山梨県': 'yamanashi', '長野県': 'nagano', '岐阜県': 'gifu', '静岡県': 'shizuoka',
  '愛知県': 'aichi', '三重県': 'mie',
  '滋賀県': 'shiga', '京都府': 'kyoto', '大阪府': 'osaka', '兵庫県': 'hyogo',
  '奈良県': 'nara', '和歌山県': 'wakayama',
  '鳥取県': 'tottori', '島根県': 'shimane', '岡山県': 'okayama', '広島県': 'hiroshima', '山口県': 'yamaguchi',
  '徳島県': 'tokushima', '香川県': 'kagawa', '愛媛県': 'ehime', '高知県': 'kochi',
  '福岡県': 'fukuoka', '佐賀県': 'saga', '長崎県': 'nagasaki', '熊本県': 'kumamoto',
  '大分県': 'oita', '宮崎県': 'miyazaki', '鹿児島県': 'kagoshima', '沖縄県': 'okinawa',
};

async function fetchCountForSlug(slug) {
  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    const res = await fetch(`https://suumo.jp/chintai/${slug}/`, {
      signal: controller.signal,
      headers: {
        'user-agent': 'Mozilla/5.0 (compatible; JapanOSINT/1.0)',
        'accept': 'text/html',
      },
    });
    clearTimeout(timer);
    if (!res.ok) return null;
    const html = await res.text();
    // Suumo prints totals like "物件数 123,456 件" or "該当件数 123,456件".
    const m = html.match(/([0-9,]{2,})\s*件/);
    if (!m) return null;
    const n = parseInt(m[1].replace(/,/g, ''), 10);
    return Number.isFinite(n) ? n : null;
  } catch {
    return null;
  }
}

export default async function collectSuumoRentalDensity() {
  const entries = Object.entries(SUUMO_SLUGS);
  const features = [];
  let liveHits = 0;

  for (const [prefJa, slug] of entries) {
    const centroid = PREFECTURE_CENTROID[prefJa];
    if (!centroid) continue;
    // Sequential polling with a tiny gap to stay under Suumo's radar.
    const count = await fetchCountForSlug(slug);
    if (count != null) liveHits++;
    features.push({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [centroid.lon, centroid.lat] },
      properties: {
        id: `SUUMO_${slug}`,
        prefecture_ja: prefJa,
        prefecture_slug: slug,
        rental_listings: count,
        source: count != null ? 'suumo_live' : 'suumo_unavailable',
      },
    });
    await new Promise((r) => setTimeout(r, 250));
  }

  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: liveHits > 0 ? 'suumo_scrape' : 'suumo_offline',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      live_hits: liveHits,
      cadence_hint: 'Run at most once per 24 h — respect Suumo robots.txt',
      description: 'Suumo rental-listing counts per prefecture (density proxy)',
    },
  };
}

File: native/collectors/classifieds/sources/suumo_rental_density.c
/* collectors/classifieds/sources/suumo_rental_density.c
 * Port of server/src/collectors/suumoRentalDensity.js.
 * For each of 47 prefectures: GET https://suumo.jp/chintai/<slug>/ , parse
 * first /([0-9,]{2,})\s*件/ → rental_listings count, pin at pref centroid.
 * ALWAYS emits one Feature per pref (count null → source 'suumo_unavailable'),
 * so 47 rows regardless of upstream — not a SEED fallback (HARD RULE 8 N/A).
 * Feature props (exact JS order): id, prefecture_ja, prefecture_slug,
 *   rental_listings, source.  id="SUUMO_<slug>" → uid via NATIVE_ID "id". */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../lib/geojson.h"
#include "../../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct pe { const char *ja, *slug; double lat, lon; };

/* SUUMO_SLUGS iteration order (JS Object insertion order) with the matching
 * PREFECTURE_CENTROID lat/lon. */
static const struct pe PREFS[] = {
  {"北海道","hokkaido",43.0642,141.3469},{"青森県","aomori",40.8243,140.7400},
  {"岩手県","iwate",39.7036,141.1527},{"宮城県","miyagi",38.2688,140.8721},
  {"秋田県","akita",39.7186,140.1024},{"山形県","yamagata",38.2404,140.3634},
  {"福島県","fukushima",37.7503,140.4676},{"茨城県","ibaraki",36.3418,140.4468},
  {"栃木県","tochigi",36.5657,139.8836},{"群馬県","gumma",36.3912,139.0608},
  {"埼玉県","saitama",35.8570,139.6489},{"千葉県","chiba",35.6047,140.1233},
  {"東京都","tokyo",35.6895,139.6917},{"神奈川県","kanagawa",35.4478,139.6425},
  {"新潟県","nigata",37.9161,139.0364},{"富山県","toyama",36.6953,137.2113},
  {"石川県","ishikawa",36.5946,136.6256},{"福井県","fukui",36.0652,136.2216},
  {"山梨県","yamanashi",35.6640,138.5683},{"長野県","nagano",36.6513,138.1811},
  {"岐阜県","gifu",35.3912,136.7223},{"静岡県","shizuoka",34.9770,138.3830},
  {"愛知県","aichi",35.1802,136.9066},{"三重県","mie",34.7303,136.5086},
  {"滋賀県","shiga",35.0045,135.8686},{"京都府","kyoto",35.0212,135.7556},
  {"大阪府","osaka",34.6863,135.5200},{"兵庫県","hyogo",34.6913,135.1830},
  {"奈良県","nara",34.6851,135.8050},{"和歌山県","wakayama",34.2261,135.1675},
  {"鳥取県","tottori",35.5036,134.2378},{"島根県","shimane",35.4723,133.0505},
  {"岡山県","okayama",34.6618,133.9350},{"広島県","hiroshima",34.3963,132.4596},
  {"山口県","yamaguchi",34.1859,131.4707},{"徳島県","tokushima",34.0657,134.5592},
  {"香川県","kagawa",34.3401,134.0434},{"愛媛県","ehime",33.8416,132.7657},
  {"高知県","kochi",33.5597,133.5311},{"福岡県","fukuoka",33.6064,130.4181},
  {"佐賀県","saga",33.2494,130.2988},{"長崎県","nagasaki",32.7503,129.8779},
  {"熊本県","kumamoto",32.7898,130.7417},{"大分県","oita",33.2382,131.6126},
  {"宮崎県","miyazaki",31.9111,131.4239},{"鹿児島県","kagoshima",31.5602,130.5581},
  {"沖縄県","okinawa",26.2124,127.6809},
};
#define NPREF ((int)(sizeof PREFS / sizeof PREFS[0]))

/* JS: html.match(/([0-9,]{2,})\s*件/); n = parseInt(g1.replace(/,/,''),10);
 * Number.isFinite(n)?n:null. The {2,} means the (digit|comma) run is >=2 chars.
 * Returns 1 + sets *out if a match yields a finite integer, else 0. */
static int parse_count(const char *html, long *out) {
  const char *p = html;
  while (*p) {
    if ((*p >= '0' && *p <= '9')) {
      const char *s = p;
      while (*p && ((*p >= '0' && *p <= '9') || *p == ',')) p++;
      int runlen = (int)(p - s);
      const char *u = p;
      while (*u == ' ' || *u == '\t' || *u == '\n' || *u == '\r' || *u == '\f' || *u == '\v') u++;
      /* "件" = E4 BB B6 */
      if (runlen >= 2 && (unsigned char)u[0] == 0xE4
          && (unsigned char)u[1] == 0xBB && (unsigned char)u[2] == 0xB6) {
        long v = 0; int any = 0;
        for (const char *d = s; d < p; d++) {
          if (*d == ',') continue;
          v = v * 10 + (*d - '0'); any = 1;
        }
        if (any) { *out = v; return 1; }
      }
    } else {
      p++;
    }
  }
  return 0;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  cJSON *features = cJSON_CreateArray();
  for (int i = 0; i < NPREF; i++) {
    char url[128];
    snprintf(url, sizeof url, "https://suumo.jp/chintai/%s/", PREFS[i].slug);
    char *html = feed_get_text(ctx->http, url, 15000);

    long count = 0; int have = 0;
    if (html) { have = parse_count(html, &count); free(html); }

    cJSON *f = cJSON_CreateObject();
    cJSON_AddStringToObject(f, "type", "Feature");
    cJSON *g = cJSON_CreateObject();
    cJSON_AddStringToObject(g, "type", "Point");
    cJSON *co = cJSON_CreateArray();
    cJSON_AddItemToArray(co, cJSON_CreateNumber(PREFS[i].lon));
    cJSON_AddItemToArray(co, cJSON_CreateNumber(PREFS[i].lat));
    cJSON_AddItemToObject(g, "coordinates", co);
    cJSON_AddItemToObject(f, "geometry", g);

    cJSON *p = cJSON_CreateObject();   /* EXACT JS key order */
    char id[64]; snprintf(id, sizeof id, "SUUMO_%s", PREFS[i].slug);
    cJSON_AddStringToObject(p, "id", id);
    cJSON_AddStringToObject(p, "prefecture_ja", PREFS[i].ja);
    cJSON_AddStringToObject(p, "prefecture_slug", PREFS[i].slug);
    cJSON_AddItemToObject(p, "rental_listings",
      have ? cJSON_CreateNumber((double)count) : cJSON_CreateNull());
    cJSON_AddStringToObject(p, "source",
      have ? "suumo_live" : "suumo_unavailable");
    cJSON_AddItemToObject(f, "properties", p);
    cJSON_AddItemToArray(features, f);
  }

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[suumo-rental-density] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def suumo_rental_density_def = {
  .id = "suumo-rental-density", .collector = "classifieds",
  .name = "Suumo Rental Density", .name_ja = "Suumo 賃貸物件数",
   .update_interval_sec = 86400, .run = run,
};
REGISTER_SOURCE(suumo_rental_density_def)
